src/wallet/asyncrpcoperation_shieldcoinbase.cpp

```diff
@@ -43,95 +41,47 @@
 
 using namespace libzcash;
 
-static int find_output(UniValue obj, int n) {
-    UniValue outputMapValue = find_value(obj, "outputmap");
-    if (!outputMapValue.isArray()) {
-        throw JSONRPCError(RPC_WALLET_ERROR, "Missing outputmap for JoinSplit operation");
-    }
-
-    UniValue outputMap = outputMapValue.get_array();
-    assert(outputMap.size() == ZC_NUM_JS_OUTPUTS);
-    for (size_t i = 0; i < outputMap.size(); i++) {
-        if (outputMap[i].get_int() == n) {
-            return i;
-        }
-    }
-
-    throw std::logic_error("n is not present in outputmap");
-}
-
 AsyncRPCOperation_shieldcoinbase::AsyncRPCOperation_shieldcoinbase(
-        TransactionBuilder builder,
-        CMutableTransaction contextualTx,
-        std::vector<ShieldCoinbaseUTXO> inputs,
+        WalletTxBuilder builder,
+        ZTXOSelector ztxoSelector,
         PaymentAddress toAddress,
+        TransactionStrategy strategy,
+        int nUTXOLimit,
         CAmount fee,
         UniValue contextInfo) :
-        builder_(std::move(builder)), tx_(contextualTx), inputs_(inputs), fee_(fee), contextinfo_(contextInfo)
+    builder_(std::move(builder)),
+    ztxoSelector_(ztxoSelector),
+    toAddress_(toAddress),
+    strategy_(strategy),
+    nUTXOLimit_(nUTXOLimit),
+    fee_(fee),
+    contextinfo_(contextInfo)
 {
-    assert(contextualTx.nVersion >= 2);  // transaction format version must support vJoinSplit
-
-    if (fee < 0 || fee > MAX_MONEY) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, "Fee is out of range");
-    }
-
-    if (inputs.size() == 0) {
-        throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Empty inputs");
-    }
-
-    //  Check the destination address is valid for this network i.e. not testnet being used on mainnet
-    examine(toAddress, match {
-        [&](CKeyID addr) {
-            throw JSONRPCError(RPC_VERIFY_REJECTED, "Cannot shield coinbase output to a p2pkh address.");
-        },
-        [&](CScriptID addr) {
-            throw JSONRPCError(RPC_VERIFY_REJECTED, "Cannot shield coinbase output to a p2sh address.");
-        },
-        [&](libzcash::SaplingPaymentAddress addr) {
-            tozaddr_ = addr;
-        },
-        [&](libzcash::SproutPaymentAddress addr) {
-            tozaddr_ = addr;
-        },
-        [&](libzcash::UnifiedAddress addr) {
-            tozaddr_ = addr;
-        }
-    });
+    assert(MoneyRange(fee));
```

Comment:
Yes – `AmountFromValue` which is used across the RPC to create `CAmount` from `UniValue` throws on `!MoneyRange` before we ever get here.

However, I would like to make the constructor for `CAmount` private and add a convenience constructor that returns optional or expected on `< -MAX_MONEY` or `> MAX_MONEY`, and then something like `class CMoney : CAmount` with a similar convenience constructor that tightens the check to `< 0`. Then we can know that values of these types are always in the correct range, avoiding asserts and checks elsewhere.

Reply:
Oh, `CAmount` isn’t a class, of course. I was thinking of `CFeeRate`. So, nvm.

Reply:
+1 on adding wrappers for signed and unsigned amounts, even if they have to be separate from CAmount so that we can migrate gradually. We'll need to do that anyway for ZSAs.